x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+		if err != nil {
+			if os.IsNotExist(err) {
+				// Skip - file might have been rotated out
+				r.log.Debugf("File %v does not exist anymore.", path)
+				continue
+			} else {
+				return nil, errors.Wrapf(err, "unexpected error when looking up file %v", path)
+			}
+		} else if fileInfo.Sys() == nil {
+			return nil, fmt.Errorf("empty stat result for file %v", path)
+		}
+
+		if r.log.IsDebug() {
+			inode := Inode(fileInfo.Sys().(*syscall.Stat_t).Ino)
+			if _, isKnownFile := r.fileRecords[inode]; !isKnownFile {
+				r.log.Debugf("Found new file: %v (size=%v)", path, fileInfo.Size())
+			}
+		}
+
+		fileInfos[path] = fileInfo
+	}
+
+	return fileInfos, nil
+}
+
+func (r *UtmpFileReader) updateFileRecord(inode Inode, size int64, utmpRecords *[]Utmp) {
+	newFileRecord := FileRecord{
+		Inode: inode,
+		Size:  size,
+	}
+
+	if utmpRecords != nil && len(*utmpRecords) > 0 {
+		newFileRecord.LastUtmp = (*utmpRecords)[len(*utmpRecords)-1]
+	} else {
+		oldFileRecord, found := r.fileRecords[inode]
+		if found {
+			newFileRecord.LastUtmp = oldFileRecord.LastUtmp
+		}
+	}
+
+	r.fileRecords[inode] = newFileRecord
+}
+
+// ReadAfter reads a UTMP formatted file (usually /var/log/wtmp*)
+// and returns the records after the provided last known record.
+// If record is nil, it returns all records in the file.
+func (r *UtmpFileReader) readAfter(path string, lastKnownRecord *Utmp) ([]Utmp, error) {
+	cs := C.CString(path)
+	defer C.free(unsafe.Pointer(cs))
+
+	success, err := C.utmpname(cs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error selecting UTMP file %v", path)
+	}
+	if success != 0 {
+		return nil, fmt.Errorf("selecting UTMP file %v failed", path)
+	}
+
+	C.setutent()
+	defer C.endutent()
+
+	reachedNewRecords := (lastKnownRecord == nil)
+	var utmpRecords []Utmp
+	for {
+		utmpC, err := C.getutent()
+		if err != nil {
+			return nil, errors.Wrapf(err, "error getting entry in UTMP file %v", path)
+		}
+
+		if utmpC != nil {
+			utmp := newUtmp(utmpC)
+
+			if reachedNewRecords {
+				r.log.Debugf("utmp: (ut_type=%d, ut_pid=%d, ut_line=%v, ut_user=%v, ut_host=%v, ut_tv.tv_sec=%v, ut_addr_v6=%v)",
+					utmp.UtType, utmp.UtPid, utmp.UtLine, utmp.UtUser, utmp.UtHost, utmp.UtTv, utmp.UtAddrV6)
+
+				utmpRecords = append(utmpRecords, utmp)
+			}
+
+			if !reachedNewRecords && lastKnownRecord != nil && reflect.DeepEqual(utmp, *lastKnownRecord) {
+				reachedNewRecords = true
+			}
+		} else {
+			// Eventually, we have read all UTMP records in the file.
+
+			if !reachedNewRecords && lastKnownRecord != nil {
+				// For some reason, this file did not contain the saved record.
+				// This might be a sign of a highly unlikely inode reuse -
+				// or of something more nefarious. We go back to the beginning and
+				// read the whole file this time.
+				r.log.Warnf("Unexpectedly, the file %v did not contain the saved login record %v - reading whole file.",
+					path, *lastKnownRecord)
+
+				return r.readAfter(path, nil)
```

Comment:
Looks like we are ignoring all messages read so far if we end up here (e.g. we read while copytruncate like logrotation happens). We still want to publish known records? Alternatively we can just error and return here, but rely on fetch to reset the file offset and read new contents the next time Fetch is executed.

Reply:
But can `wtmp` be copy-truncated? I thought it's guaranteed to be renamed with moves. I wouldn't want to bring all the complexity of Filebeat's corner cases in here :)

Reply:
Rename with move should be ok. But them we don't get into this code path.

Instead of having a tail recursive call we could use a for loop and reset some state and use `continue` in order to jump to beginning of the reader and not drop events. But then it's an edge case that normally should not occur.

Reply:
The code is now using offsets and if there is some problem in jumping to the stored offset it will jump back to the beginning and re-read the file. This is maybe a bit overzealous, but I think at least for now I'd prefer it if events were sent twice rather than not at all. If there are any problems we can always change.